Added Quad9 analyzer and refactored DNS analyzers

drosetti opened the pull request:
### DNS analyzers have a folder hierarchy:
```
dns
|
|_____dns_malicious_detectors
|_____dns_resolvers
|_____dns_responses.py
```

Folder explainations:
* `dns_malicious_detectors`: modules in this folder have analyzers that check in the external provider databases if the domain inside the DNS query is flagged as malicious
* `dns_resolvers`: modules in this folder have analyzers that perform a DNS query and return the resolutions
* `dns_responses.py`: modules with utilities functions. Using these functions every analyzers in one folder return the same data.
    * Analyzers in `dns_malicious_detectors` return a dict: 
    ```{"observable": <observable_name>, "malicious": True/False}```
    * Analyzers in `dns_resolvers` return a dict (resolutions is empty for NXDOMAIN):
    ```
   {
       "observable": <observable_name>,
       "resolutions": [
            {
             "TTL": <int>,
             "data": <related_data>,
             "name": <domain_name>,
             "type": <DNS_query_type>
            }
       ]
    }
    ```

### Analyzers 
The old DNS analyzers in `active_dns.py` have been moved:
* `__doh_google` (function) -> `dns_resolvers/google_dns_resolver` (module)
* `__doh_cloudflare` (function) -> `dns_resolvers/cloudflare_dns_resolver` (module)
* `__doh_cloudflare_malware` (function) -> `dns_malicious_detectors/cloudflare_malicious_detector` (module)
* `__classic_dns` (function) -> `dns_resolvers/classic_dns_resolver` (module)
*  'googlesf' (module) -> `dns_malicious_detectors/googlesf` (module)

Quad9 support both malicious detector (`dns_malicious_detectors/quad9`) and DNS resolver (`dns_resolvers/quad9_dns_resolver`)

### Added support of different DNS query types
Every `dns_resolvers` can filter DNS queries with `query_type` additional param in the `analyzer_config`:
```
"additional_config_params": {
      "query_type": "A"
    }
``` 

### Modified analyzers name
The names of `active_dns` analyzer has been changed in the `analyzer_config` file and in the GUI:
* ActiveDNS_Classic -> Classic_DNS
* ActiveDNS_CloudFlare -> CloudFlare_DNS
* ActiveDNS_CloudFlare_Malware -> CloudFlare_Malicious_Detector
* ActiveDNS_Google -> Google_DNS

mlodic commented on api_app/script_analyzers/observable_analyzers/dns/dns_malicious_detectors/quad9_malicious_detector.py:
```diff
@@ -0,0 +1,78 @@
+
+
+class Quad9MaliciousDetector(classes.ObservableAnalyzer):
+    """Retrieve malicious domain in Quad9 DB.
```
Just a reformat of the message:
```
Check if a domain is malicious by Quad9 public resolver.
Quad9 does not answer in the case a malicious domain is queried.
However, we need to perform another check to understand if that domain was blocked by the resolver or if it just does not exist.
So we perform one request to Quad9 and another one to Google.
In the case of empty response from Quad9 and a not-empty response from Google,
we can guess that the domain was in the Quad9 blacklist.
```

mlodic commented on api_app/script_analyzers/observable_analyzers/dns/dns_malicious_detectors/quad9_malicious_detector.py:
```diff
@@ -0,0 +1,78 @@
+
+    def _quad9_dns_query(self, observable):
+        """Perform a DNS query with Quad9 service, return True if Quad9 answer the
+        DNS query.
```
...with a non-empty response

mlodic commented on tests/utils.py:
```diff
@@ -185,6 +227,36 @@ def test_fortiguard(self, mock_get=None, mock_post=None):
+
+    def test_quad9_malicious_detector(self, mock_get=None, mock_post=None):
+        report = quad9_malicious_detector.Quad9MaliciousDetector(
+            "GoogleDNSResolver",
```
bad analyzer name

Eshaan7 commented:
To resolve the merge conflicts, remove the function you defined in `tasks.py` and add the `soft_time_limit` as an attribute to the analyzer config you made for quad9 in `analyzer_config.json`.

drosetti commented:
Hi Eshaan, 

Thanks for you support, i will follow your advice and update my code.

drosetti commented:
I have merged the conflicts as you suggest, I have also added "docs" folder to `.flake8` in order to ignore annoying flake8 message for docs theme import. 

I'have also modified `intel_owl.tasks.run_analyzer()`, now support nested packages like `dns`.

mlodic commented:
Well done, merged! thank you!
